compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -2355,6 +2358,58 @@ public void visit(BLangTupleDestructure tupleDeStmt, AnalyzerData data) {
         if (type.tag != TypeTags.SEMANTIC_ERROR) {
             checkTupleVarRefEquivalency(tupleDeStmt.pos, tupleDeStmt.varRef,
                                         tupleDeStmt.expr.getBType(), tupleDeStmt.expr.pos, data);
+            if (tupleDeStmt.expr.getKind() == LIST_CONSTRUCTOR_EXPR) {
+                analyzeDestructuringWithListBindingPattern(tupleDeStmt.varRef,
+                        (BLangListConstructorExpr) tupleDeStmt.expr);
+            }
+        }
+    }
+
+    private void analyzeDestructuringWithListBindingPattern(BLangTupleVarRef tupleVarRef,
+                                                            BLangListConstructorExpr listExpr) {
+        List<BLangExpression> exprs = listExpr.exprs;
+        List<BLangExpression> tupleVars = tupleVarRef.expressions;
+
+        int nonRestTupleVarIndex = 0;
+        int noOfNonRestVars = tupleVars.size();
+
+        for (BLangExpression expr: exprs) {
+            int remainingNonRestVars = noOfNonRestVars - nonRestTupleVarIndex;
+
+            if (remainingNonRestVars == 0) {
+                return;
+            }
+
+            if (expr.getKind() == NodeKind.LIST_CONSTRUCTOR_SPREAD_OP) {
+                BType spreadOpType = ((BLangListConstructorSpreadOpExpr) expr).expr.getBType();
+                int noOfSpreadMembers = switch (spreadOpType.tag) {
+                    case TypeTags.ARRAY -> {
+                        BArrayType spreadArrayType = (BArrayType) spreadOpType;
+                        yield spreadArrayType.size;
+                    }
+                    case TypeTags.TUPLE -> {
+                        BTupleType spreadTupleType = (BTupleType) spreadOpType;
+                        yield spreadTupleType.getTupleTypes().size();
+                    }
+                    default -> 0;
+                };
+
+                // remainingNonRestVars < noOfSpreadMembers case is ignored since it is already invalid scenario
+
+                nonRestTupleVarIndex += noOfSpreadMembers;
+                continue;
+            }
+
+            BLangExpression tupleMemberVar = tupleVars.get(nonRestTupleVarIndex++);
+            if (expr.getKind() == LIST_CONSTRUCTOR_EXPR && tupleMemberVar.getKind() == TUPLE_VARIABLE_REF) {
```

Comment:
Exteact `expr.getKind()` (used above also) and `tupleMemberVar.getKind()` out to variables.